### ISSUE TITLE
parallax: Limit parallax scroll effect

### ISSUE THREAD

**Issue by afilbert** [ampproject/amphtml#11019]
Looks like the `amp-fx-parallax` component utilizes total scroll height of the page, and calculates the scroll position of whatever target element based on a ratio of that height. This means that if you have sufficient content above where you want the parallax effect, you could end up not even seeing either the effect or the targeted parallax content. Please see the image below for an example where the content is scrolled up and completely obscured.

## How do we reproduce the issue?

1. Open up the parallax component example: https://ampbyexample.com/components/amp-fx-parallax/
2. Scroll down until you can see the image of the fawn. By that time, you can start to see the end of the `Title 20` content scrolling up beyond the bottom of the image.

Given more content above the demo, if adding a large spacer or more paragraphs of text - for example, the issue would be even more pronounced; eventually making it so you can't even see the scrolled content at all by the time that section of the page becomes viewable.

Example:

<img width="1076" alt="screen shot 2017-08-21 at 7 27 47 pm" src="https://user-images.githubusercontent.com/523157/29543874-f964a2ea-86a9-11e7-9491-db2454a3d576.png">

*Use-case:*

I'm working on a template for the [ampstart](https://github.com/ampproject/ampstart) project where we're trying to showcase the `amp-fx-parallax` component using a background image. That image should scroll up slightly the further the user scrolls down. The more content I add to the template, however, the more pronounced the effect. Depending on the width of the viewport, the page gets quite long, and eventually the background image disappears entirely:

Before:

<img width="632" alt="screen shot 2017-08-21 at 8 01 26 pm" src="https://user-images.githubusercontent.com/523157/29544115-94513f6a-86ab-11e7-891f-bf90b52b4fe2.png">

After:

<img width="456" alt="screen shot 2017-08-21 at 7 59 27 pm" src="https://user-images.githubusercontent.com/523157/29544126-a785f378-86ab-11e7-839a-ab6d0b414da1.png">

The only difference between those screencaps being the screen width. This is true even with a factor of 1.1.

Using a taller image is one workaround, but since the template will utilize `amp-accordion` sections, along with a responsive design, it will have arbitrary page height for the end user.

## Outcomes

I'd like to see a parameter or setting in addition to the `amp-fx-parallax=<factor>` that controls the extent to which targeted content scrolls up or down relative to the page scroll position. Something like a total pixel amount before the targeted content stops, for example. Ideally, it would make sense in our use-case to prevent the bottom of the `amp-image` from scrolling up past the bottom of its container.

## Affected Browsers/AMP Version

Browsers:

Looks like any browser that supports the parallax component is affected. I've tested latest (as of 8/21/17) versions of Safari, Firefox, and Chrome web browsers.

AMP Version: 

Here are the includes in my ampstart project:

`https://cdn.ampproject.org/v0.js`
`https://cdn.ampproject.org/v0/amp-fx-parallax-0.1.js`

**Comment by afilbert** [comment]
Here's my code:

_Markup_

```
<div class="land-see-subscribe center items-center max-width-4 mx-auto m0 mb3">
      <div class="land-see-subscribe-form">
        {{> ../../components/form/form.snip.html}}
      </div>
      <amp-img amp-fx-parallax="1.1" alt="A view of the sea, see?"
        layout="fixed-height"
        height="679"
        src="../../img/land-see/content/khatam-tadayon-107740.jpg">
      </amp-img>
    </div>
```

_CSS_

```
.land-see-subscribe {
  position: relative;
  height: 310px;
  overflow: hidden;
  color: var(--color-primary);
}

amp-img.cover img {
  object-fit: cover;
}
```

**Comment by afilbert** [comment]
After pondering this some more, it occurred to me that limiting the travel height of targeted `amp-fx-parallax` content isn't enough. That's desirable, sure, but that limitation could just mean that the bottom of the target content could already be reached by the time things scroll into view (depending on the height of the page). Being able to delay scroll animation until within a certain range of the viewport may also be something to consider.

**Comment by aghassemi** [comment]
Hi @afilbert 

I have been working on a low-level building block ( #10818 ) to enable scrollbound and visibility-based FX (including parallax) in AMP. It is ready for experimentation and I believe it will handle your use-case.

Please take a look at the documentation here: https://github.com/ampproject/amphtml/blob/master/extensions/amp-position-observer/amp-position-observer.md 

I have a Parallax Scene sample using it here as well: https://s.codepen.io/aghassemi/debug/dzKGWy
  
Code for the sample (and other samples using this new approach) can be found here:  https://codepen.io/collection/nMJYrv/ 

We most likely will keep `amp-fx-parallax` but purely a syntax sugar for ease of us, behind the scene it will change to use this new approach.

I would love to hear your feedback/bug reports on this.

**Comment by afilbert** [comment]
Thanks, @aghassemi. I was able to implement the position observer. You were right, it worked great for my use-case. Your Parallax Scene example really helped. Here's a link to the Land + See implementation (footer subscribe section):

https://landseeampstart-exp.firebaseapp.com/templates/land-see/land-see.amp

I'm likely going to implement similar with additional background and accent elements, as well. I'll definitely let you know if I have any feedback.

**Comment by aghassemi** [comment]
@afilbert great, cool page!

**Comment by afilbert** [comment]
@aghassemi Struggling with applying a general animation to many/dynamically-generated "post" items. Not sure if moved beyond intended use-case, not implementing correctly, or dealing with potential bugs. I've created an issue, here:

https://github.com/ampproject/amphtml/issues/11167

I believe this issue, however, is resolved by #10818. I can close, if needed...

**Comment by aghassemi** [comment]
Closing per https://github.com/ampproject/amphtml/issues/11167#issuecomment-328668874
